eras/dijkstra/impl/src/Cardano/Ledger/Dijkstra/TxBody.hs

```diff
@@ -180,82 +258,109 @@ basicDijkstraTxBodyRaw =
           (null . unWithdrawals)
-          (\x tx -> tx {dtbrWithdrawals = x})
-          From
-      bodyFields 7 = ofield (\x tx -> tx {dtbrAuxDataHash = x}) From
-      bodyFields 8 =
-        ofield
-          (\x tx -> tx {dtbrVldt = (dtbrVldt tx) {invalidBefore = x}})
+          (\x tx -> tx & withdrawalsDijkstraTxBodyRawL .~ x)
           From
+      bodyFields 7 = ofield (\x tx -> tx & auxDataHashDijkstraTxBodyRawL .~ x) From
+      bodyFields n@8 =
+        withSTxBothLevels @l $ \case
+          STopTx ->
+            ofield
+              (\x tx -> tx & vldtDijkstraTxBodyRawL .~ (dtbrVldt tx) {invalidBefore = x})
+              From
+          SSubTx ->
+            invalidField n
       bodyFields 9 =
         fieldGuarded
           (emptyFailure "Mint" "non-empty")
           (== mempty)
-          (\x tx -> tx {dtbrMint = x})
-          From
-      bodyFields 11 = ofield (\x tx -> tx {dtbrScriptIntegrityHash = x}) From
-      bodyFields 13 =
-        fieldGuarded
-          (emptyFailure "Collateral Inputs" "non-empty")
-          null
-          (\x tx -> tx {dtbrCollateralInputs = x})
+          (\x tx -> tx & mintDijkstraTxBodyRawL .~ x)
           From
+      bodyFields 11 = ofield (\x tx -> tx & scriptIntegrityHashDijkstraTxBodyRawL .~ x) From
+      bodyFields n@13 =
+        withSTxBothLevels @l $ \case
+          STopTx ->
+            fieldGuarded
+              (emptyFailure "Collateral Inputs" "non-empty")
+              null
+              (\x tx -> tx & collateralInputsDijkstraTxBodyRawL .~ x)
+              From
+          SSubTx -> invalidField n
       bodyFields 14 =
         ofield
-          (\x tx -> tx {dtbrGuards = fromSMaybe mempty x})
+          (\x tx -> tx & guardsDijkstraTxBodyRawL .~ fromSMaybe mempty x)
           (D decodeGuards)
-      bodyFields 15 = ofield (\x tx -> tx {dtbrNetworkId = x}) From
-      bodyFields 16 = ofield (\x tx -> tx {dtbrCollateralReturn = x}) From
-      bodyFields 17 = ofield (\x tx -> tx {dtbrTotalCollateral = x}) From
+      bodyFields 15 = ofield (\x tx -> tx & networkIdDijkstraTxBodyRawL .~ x) From
+      bodyFields n@16 =
+        withSTxBothLevels @l $ \case
+          STopTx -> ofield (\x tx -> tx & collateralReturnDijkstraTxBodyRawL .~ x) From
+          SSubTx -> invalidField n
+      bodyFields n@17 =
+        withSTxBothLevels @l $ \case
+          STopTx -> ofield (\x tx -> tx & totalCollateralDijkstraTxBodyRawL .~ x) From
+          SSubTx -> invalidField n
       bodyFields 18 =
         fieldGuarded
           (emptyFailure "Reference Inputs" "non-empty")
           null
-          (\x tx -> tx {dtbrReferenceInputs = x})
+          (\x tx -> tx & referenceInputsDijkstraTxBodyRawL .~ x)
           From
       bodyFields 19 =
         fieldGuarded
           (emptyFailure "VotingProcedures" "non-empty")
           (null . unVotingProcedures)
-          (\x tx -> tx {dtbrVotingProcedures = x})
+          (\x tx -> tx & votingProceduresDijkstraTxBodyRawL .~ x)
```

Comment:
Some eta reduction? (If you like the suggestion, please apply it to all other places like this)
```suggestion
          (\x -> votingProceduresDijkstraTxBodyRawL .~ x)
```